File: dodam-application/dodam-rest-api/src/main/java/b1nd/dodam/restapi/support/meal/MealUseCase.java
package b1nd.dodam.restapi.support.meal;

import b1nd.dodam.neis.meal.client.data.Meal;
import lombok.RequiredArgsConstructor;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
@RequiredArgsConstructor
@Transactional
public class MealUseCase {

    @Cacheable(value = "meal-day", key = "#meal.date.format(T(java.time.format.DateTimeFormatter).ofPattern('yyyy-M-d'))")
    public Meal saveMeal(Meal meal) {
        return meal;
    }
}

File: services/service-neis/src/main/kotlin/com/b1nd/dodamdodam/neis/application/meal/MealUseCase.kt
package com.b1nd.dodamdodam.neis.application.meal

import com.b1nd.dodamdodam.core.common.data.Response
import com.b1nd.dodamdodam.neis.application.meal.data.response.MealResponse
import com.b1nd.dodamdodam.neis.application.meal.data.toResponse
import com.b1nd.dodamdodam.neis.domain.meal.service.MealService
import com.b1nd.dodamdodam.neis.infrastructure.neis.NeisClient
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.time.YearMonth

@Component
@Transactional(readOnly = true)
class MealUseCase(
    private val mealService: MealService,
    private val neisClient: NeisClient,
) {
    fun getMealsByDate(date: LocalDate): Response<List<MealResponse>> {
        val meals = mealService.getMealsByDate(date)
        return Response.ok("급식을 조회했어요.", meals.map { it.toResponse() })
    }

    @Transactional
    fun syncMeals(yearMonth: YearMonth): Response<Any> {
        val meals = neisClient.fetchMonthlyMeals(yearMonth)
        meals.forEach { meal ->
            mealService.saveOrUpdate(meal.date, meal.mealType, meal.calorie, meal.menus)
        }
        return Response.ok("${yearMonth} 급식 동기화가 완료되었어요.")
    }
}